carmm/run/workflows/react.py

```diff
@@ -548,6 +609,120 @@ def vibrate(self, atoms: Atoms, indices: list, read_only=False):
             vib.write_mode()
             return vib
 
+    @property
+    def MaceReact_Preoptimiser(self):
+        """
+        Get or set the MACE preoptimiser used in aims_optimise or ts_search with setting
+
+        Args:
+             MaceReact MaceReact Obj:
+               User-defined MaceReact.
+
+        Returns:
+             self._MaceReactor MaceReact Obj:
+               User-defined MaceReact.
+        """
+
+        return self._MaceReactor
+
+    @MaceReact_Preoptimiser.setter
+    def MaceReact_Preoptimiser(self, MaceReact):
+
+        self._MaceReactor = MaceReact
+
+    def _mace_preoptimise(self, atoms: Atoms, fmax, relax_unit_cell, optimiser, opt_kwargs = {}):
+        """
+        Internal function for preoptimisation of structures with a pre-set MACE calculator
+
+        Args:
+            atoms: Atoms object
+            fmax: flaot
+            relax_unit_cell: bool
+            optimiser: bool or optimiser class
+            opt_kwargs: dict
+        Return:
+           preopt_atoms: Atoms object
+              Atoms object with the calculator object removed
+        """
+
+        filname = self._MaceReactor.filename
+        self._MaceReactor.filename = "MACE_PREOPT_" + self.filename
+
+        preopt_atoms = self._MaceReactor.mace_optimise(atoms, fmax, restart=False, relax_unit_cell=relax_unit_cell,
+                                                       optimiser = optimiser, opt_kwargs = opt_kwargs)
+
+        self._MaceReactor.filename = filname
+
+        # Clean calculator to prevent future problems
+        preopt_atoms.calc = None
+
+        return preopt_atoms
+
+    def _mace_preoptimise_ts(self, initial, final, fmax, n, interpolation, input_check, max_steps=200):
+        """
+        Internal function for preoptimisation of NEB with a pre-set MACE calculator
+
+        Args:
+            initial: Atoms object
```

Comment:
As above, if same as somewhere else then say so?

Reply:
Added additional comments to explain, thank you!